cmd/complex/main.go

```diff
@@ -173,3 +193,35 @@ func generateRandomNumbers() *sse.Message {
 
 	return e
 }
+
+type loggerCtxKey struct{}
+
+// withLogger is an http middleware that generates a logger with request-specific fields
+// added to it and attaches it to the request context for later retrieval with getLogger().
+// This is simmilar to how existing per-request http logging libraries work, just very simplified.
+func withLogger(logger *slog.Logger) func(h http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			l := logger.With(
+
+				"UserAgent", r.UserAgent(),
+				"RemoteAddr", r.RemoteAddr,
+				"Host", r.Host,
+				"Origin", r.Header.Get("origin"),
+			)
+			r = r.WithContext(context.WithValue(r.Context(), loggerCtxKey{}, l))
+			h.ServeHTTP(w, r)
+		})
+	}
+}
+
+// getLogger retrieves the request-specific logger from a request's context. This is
+// similar to how existing per-request http logging libraries work, just very simplified.
+func getLogger(r *http.Request) (*slog.Logger, error) {
```

Comment:
I'd honestly remove the comment and take just a `context.Context`:
```suggestion
func getLogger(ctx context.Context) (*slog.Logger, error) {
```
This way we don't really have to explain anything – not only because it's self-explanatory that it takes a logger from a context, but also because this function is wrapped when initializing the server and there's already a comment there that about wrapping third party APIs.

Possibly also get rid of the error return, to keep things simple?

Reply:
✅